### ISSUE TITLE
add support for structural equality based computations

### ISSUE THREAD

**Issue by mweststrate** [mweststrate/mobservable#8]
Sometimes you have computations in views that look like

```javascript
var coords = mobservable(function() {
    return {
        x : // some computation,
        y : // some computation
    }
});
```

Now every function that depends on `coords` will re-evaluate, even if the `x` and `y` didn't change; because there is always a fresh object returned. 

It should be possible to create (computed) observables that detect changes using deep equality instead of reference equality.

**Comment by dave-handy** [comment]
If I then have:
```
var hasX = mobservable(function() {
  return !!coords.x;
});
sideEffect(function() {
  if(hasX) console.log('X is present!');
});
coords.x = 1;
coords.x = 2;
```
How many times should "X is present!" be printed? It seems it's printed twice, even though the value of hasX doesn't change.

If this ticket is implemented will that allow code to support only one print of "X is present!"?

**Comment by mweststrate** [comment]
Hi Dave,

The above case should already print only once (if you change `hasX` to `hasX()`, the parentheses can only be omitted if the view is part of an object, so for consistency sake I recommend to always make views part of an object).

However, if your view `hasX` would return a new object on each update, than it would always be seen as a change. wrapping the function in `asStructure` fixes that by deep comparing the object returned from your view with the previous value that was returned from the view.

**Comment by dave-handy** [comment]
Hmmm I must be doing something wrong. I am actually using the observable decorator, so I forgot the parentheses. I just wrote a simplified example:

```
const state = makeReactive({
  authToken: null
});

class LoginStoreTest {
  @observable get loggedIn() {
    return !!state.authToken;
  }
}

const store = new LoginStoreTest();

sideEffect(() => {
  console.log(store.loggedIn);
});

state.authToken = 'a';
state.authToken = 'b';
```

Prints:
```
false
true
true
```

fwiw I'm on mobservable 0.6.7

**Comment by mweststrate** [comment]
Thanks for the reproduction. I can reproduce the issue in 0.6.9 as well,
and it seems to be specifically happening when the decorator is used; if I
use `extendReactive(this, { loggedIn: function })` in the `LoginStoreTest`
the issue disappears. I'll look into it later today! Thanks for noticing.

**Comment by mweststrate** [comment]
Thanks for reporting this one, it was actually a regression bug, there were no longer individual views created for getters, the results where correct, but it ran more computations then needed. Should have been fixed in the just released 0.6.10 version.

**Comment by dave-handy** [comment]
Looks like it's working now (omits the second 'true'). Thanks for the quick response! This library has really cleaned up my code.

**Comment by mweststrate** [comment]
Support for structural equality has been released as part of 0.7. (Still updating the docs though)
